gravitee-am-service/src/main/java/io/gravitee/am/service/impl/UserActivityServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+            UserActivityConfiguration configuration,
+            @Lazy UserActivityRepository userActivityRepository
+    ) {
+        this.configuration = configuration;
+        this.userActivityRepository = userActivityRepository;
+    }
+
+    public boolean canSaveUserActivity() {
+        return this.configuration.isEnabled();
+    }
+
+    public long getRetentionTime() {
+        return this.configuration.getRetentionTime();
+    }
+
+    public ChronoUnit getRetentionUnit() {
+        return this.configuration.getRetentionUnit();
+    }
+
+    public Flowable<UserActivity> findByDomainAndTypeAndUserAndLimit(String domain, Type type, String userId, int limit) {
+        return userActivityRepository.findByDomainAndTypeAndKeyAndLimit(domain, type, buildKey(userId), limit);
+    }
+
+    public Completable save(
+            String domain,
+            String userId,
+            UserActivity.Type type,
+            Map<String, Object> data) {
+        final Date createdAt = new Date();
+        return Single.defer(() -> Single.just(new UserActivity()
+                                .setReferenceType(ReferenceType.DOMAIN)
+                                .setReferenceId(domain)
+                                .setUserActivityType(type)
+                                .setUserActivityKey(buildKey(userId))
+                                .setLatitude(buildLatitude(data))
+                                .setLongitude(buildLongitude(data))
+                                .setUserAgent((String) data.get(USER_AGENT))
+                                .setLoginAttempts((Integer) data.getOrDefault(LOGIN_ATTEMPTS, 0))
+                                .setCreatedAt(createdAt)
+                                .setUpdatedAt(createdAt)
+                                .setExpireAt(getExpireAtDate(createdAt))
+                        )
+                ).flatMap(userActivityRepository::create)
+                .doOnSuccess(ua -> LOGGER.debug("UserActivity with id '{}' created", ua.getId()))
+                .doOnError(err ->
+                        LOGGER.error("An unexpected error has occurred while saving UserActivity '{}'", err.getMessage(), err)
+                ).ignoreElement();
+    }
+
+    public Completable deleteByDomainAndUser(String domain, String userId) {
+        return userActivityRepository.deleteByDomainAndKey(domain, buildKey(userId)).doOnError(err ->
+                LOGGER.error("An unexpected error has occurred while deleting userActivity '{}'", err.getMessage(), err)
+        );
+    }
+
+    public Completable deleteByDomain(String domain) {
+        return userActivityRepository.deleteByDomain(domain).doOnError(err ->
+                LOGGER.error("An unexpected error has occurred while deleting userActivity '{}'", err.getMessage(), err)
+        );
+    }
+
+    private String buildKey(String userId) {
+        return computeHash(configuration.getAlgorithmKey(), userId, configuration.getSalt());
```

Comment:
is it safe to have the same salt value for each user?
Wondering if a random value generated with a seed coming from the userId is possible :thinking:

Reply:
Let's disscuss about it